### PR TITLE
Rewards

### PR DESCRIPTION
**Pull request by danispavel**

### CODE REVIEW

File: src/js/components/CheckIn/components/GoalDate/GoalDate.js [hunk 1]
```diff
@@ -11,7 +11,14 @@ import {
...
 } from '@fortawesome/free-solid-svg-icons'
 
-import firebase from '../../../Firebase'
+import {
+  BRONZE_MAX,
+  SILVER_MAX,
+  GOLD_MAX,
+  RUBY_MAX,
+  DIAMOND_MAX,
+} from 'js/var/rewardsConsts'
```
Comment (SaenkoM) [inline src/js/components/CheckIn/components/GoalDate/GoalDate.js]: it is known that `var` folder contains constants, there is no reason to name file `rewardsConsts`

File: src/js/components/CheckIn/components/GoalDate/GoalDate.js [hunk 2]
```diff
@@ -72,22 +79,41 @@ class GoalDate extends Component {
...
+      setCheckinData,
+      setAssessmentData,
+      setRewardsCheckins,
     } = this.props
     const { value } = this.state
+    const currentCheckins = checkins + 1
 
-    event.preventDefault()
-    firebase
+    const userRef = firebase
```
Comment (SaenkoM) [inline src/js/components/CheckIn/components/GoalDate/GoalDate.js]: why reassign?

File: src/js/components/CheckIn/components/GoalDate/GoalDate.js [hunk 3]
```diff
@@ -72,22 +79,41 @@ class GoalDate extends Component {
...
       .database()
       .ref(`users/${uid}`)
-      .update({
-        goalDate: new Date(value),
-      })
 
+    userRef.update({
+      goalDate: new Date(value),
```
Comment (SaenkoM) [inline src/js/components/CheckIn/components/GoalDate/GoalDate.js]: not saving a timestamp?

File: src/js/components/CheckIn/components/GoalDate/GoalDate.js [hunk 4]
```diff
@@ -72,22 +79,41 @@ class GoalDate extends Component {
...
 
+    switch (currentCheckins) {
+      case BRONZE_MAX:
+      case SILVER_MAX:
+      case GOLD_MAX:
+      case RUBY_MAX:
+      case DIAMOND_MAX: history.push('/getcheckin')
+        break
+      default: history.push('/dashboard')
+    }
```
Comment (SaenkoM) [inline src/js/components/CheckIn/components/GoalDate/GoalDate.js]: why not just use single `if` here?

File: src/js/components/CheckIn/components/Progress/Progress.js [hunk 5]
```diff
@@ -23,14 +31,38 @@ class Progress extends Component {
...
     const { progress } = this.state
-    const { setCheckinData, setAssessmentData } = this.props
+    const {
+      history,
+      diamonds,
+      setCheckinData,
+      setAssessmentData,
+      setRewardsDiamonds,
+    } = this.props
+    const uid = firebase.auth().currentUser.uid
```
Comment (SaenkoM) [inline src/js/components/CheckIn/components/Progress/Progress.js]: `const { uid } = firebase.auth().currentUser`

File: src/js/components/CheckIn/components/Progress/Progress.js [hunk 6]
```diff
@@ -23,14 +31,38 @@ class Progress extends Component {
...
+
+      switch (currentDimonds) {
+        case BRONZE_MAX:
+        case SILVER_MAX:
+        case GOLD_MAX:
+        case RUBY_MAX:
+        case DIAMOND_MAX: history.push('/getdiamond')
+          break
+        default: return
+      }
```
Comment (SaenkoM) [inline src/js/components/CheckIn/components/Progress/Progress.js]: same, unneeded switch

File: src/js/components/Rewards/LearnEarningBadges.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            How to earn and <br /> unlock badges:
+        </p>
+        <div className="teach-badge-earning-item">
+          <div className="teach-badge-earning-item-img-wrapper">
+            <img className="teach-goal-setting-step-img" alt="first badge" src={STAR_BRONZE} />
+          </div>
+          <p className="teach-badge-earning-item-text">
+            Earn star for every goal set.
+          </p>
+        </div>
```
Comment (SaenkoM) [inline src/js/components/Rewards/LearnEarningBadges.js]: deserves a separate functional component, accepts src and text

File: src/js/components/Rewards/LearnEarningBadges.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+          <p className="teach-badge-earning-item-text">
+            Locked badges will look like this.
+          </p>
+        </div>
+        <button
+          className="button cta"
+          type="submit"
+          onClick={() => { history.push('/rewards') }}
+          >Go to my badges
+        </button>
```
Comment (SaenkoM) [inline src/js/components/Rewards/LearnEarningBadges.js]: why not use `Link` here?
opening and closing arrows should have same indentation, text and nested elements on the next line:
```
<button
  ...
>
  Go to my badges
</button

File: src/js/components/Rewards/LearnEarningBadges.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+      </div>
+    </div>
+  )
+}
+LearnEarningBadges.propTypes = {
+  history: PropTypes.object.isRequired,
+}
+
+export default LearnEarningBadges
```
Comment (SaenkoM) [inline src/js/components/Rewards/LearnEarningBadges.js]: eslint doesn't complain about empty line at the end of file?

File: src/js/components/Rewards/LearnGoalSetting/LearnGoalSetting.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+import Step2 from './components/Step2'
+import Step3 from './components/Step3'
+import Step4 from './components/Step4'
+
+class TeachGoalSetting extends Component {
+  constructor(props) {
+    super(props)
+    this.state = { step: 0 }
+    this.setStep = this.setStep.bind(this)
+  }
```
Comment (SaenkoM) [inline src/js/components/Rewards/LearnGoalSetting/LearnGoalSetting.js]: don't use `bind` to bind context, just use arrow functions, constructor is redundant

File: src/js/components/Rewards/LearnGoalSetting/LearnGoalSetting.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+  componentWillUnmount() {
+    const { incBadges } = this.props
+
+    incBadges()
```
Comment (SaenkoM) [inline src/js/components/Rewards/LearnGoalSetting/LearnGoalSetting.js]: it isn't clear what function does by name

File: src/js/components/Rewards/LearnGoalSetting/LearnGoalSetting.js [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    const { name, history, setRewardsThumbsups } = this.props
+    const { step } = this.state
+
+    return (
+      <div className="teach-goal-setting-container">
+        {step === 0 && (<Step0 setStep={this.setStep} name={name} />)}
+        {step === 1 && (<Step1 setStep={this.setStep} />)}
+        {step === 2 && (<Step2 setStep={this.setStep} />)}
+        {step === 3 && (<Step3 setStep={this.setStep} />)}
+        {step === 4 && (<Step4 history={history} setRewardsThumbsups={setRewardsThumbsups} />)}
```
Comment (SaenkoM) [inline src/js/components/Rewards/LearnGoalSetting/LearnGoalSetting.js]: `history` should be passed down by router, it doesn't?

File: src/js/components/Rewards/LearnGoalSetting/components/Step0.js [hunk 13]
```diff
@@ -0,0 +1,35 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+
+const Step0 = (props) => {
+  const { name, setStep } = props
+
+  return (
+
```
Comment (SaenkoM) [inline src/js/components/Rewards/LearnGoalSetting/components/Step0.js]: unneeded empty line

File: src/js/components/Rewards/LearnGoalSetting/components/Step0.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+            We&apos;re here to
+            help you set
+            goals.
+      </p>
+      <button
+        className="button cta"
+        type="submit"
+        onClick={() => setStep(1)}
+        >Next
+      </button>
```
Comment (SaenkoM) [inline src/js/components/Rewards/LearnGoalSetting/components/Step0.js]: indentation

File: src/js/components/Rewards/LearnGoalSetting/components/Step3.js [hunk 15]
```diff
@@ -0,0 +1,39 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+
+const Step3 = (props) => {
```
Comment (SaenkoM) [inline src/js/components/Rewards/LearnGoalSetting/components/Step3.js]: steps a very similar, why not make it one customizable component?

File: src/js/components/Rewards/badgeGetting/GetCompletedBadge/GetCompletedBadge.js [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+  SILVER_MAX,
+  GOLD_MAX,
+  RUBY_MAX,
+  DIAMOND_MAX,
+} from 'js/var/rewardsConsts'
+import firebase from 'js/components/Firebase'
+import { requireAuth } from 'js/components/Auth'
+
+class GetCompletedBadge extends Component {
+  componentDidMount =() => {
```
Comment (SaenkoM) [inline src/js/components/Rewards/badgeGetting/GetCompletedBadge/GetCompletedBadge.js]: missing space

File: src/js/components/Rewards/badgeGetting/GetCompletedBadge/GetCompletedBadge.js [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+    this.reset()
+  }
+
+  componentWillUnmount() {
+    const { incBadges } = this.props
+
+    incBadges()
+  }
+
+  reset = () => {
```
Comment (SaenkoM) [inline src/js/components/Rewards/badgeGetting/GetCompletedBadge/GetCompletedBadge.js]: i've seen that function (or the very similar one) somewhere, maybe it is possible to reuse

File: src/js/components/Rewards/components/ScrollMenu.js [hunk 18]
```diff
@@ -0,0 +1,25 @@
+import React from 'react'
+import Proptypes from 'prop-types'
+import ScrollMenu from 'react-horizontal-scrolling-menu'
+
+const CustomScrollMenu = ({ data, selected, onSelect }) => {
+  const ArrowLeft = <p className="arrow-prev" onClick={onSelect}><i className="left" /></p>
```
Comment (SaenkoM) [inline src/js/components/Rewards/components/ScrollMenu.js]: empty tags can be used  as `<p />`

File: src/js/components/Rewards/components/ScrollMenuRewardsItem.js [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+  imgAlt: PropTypes.string,
+  maxCount: PropTypes.number.isRequired,
+  className: PropTypes.string.isRequired,
+}
+
+RewardsItem.defaultProps = {
+  imgAlt: 'reward',
+}
+
+export default RewardsItem
```
Comment (SaenkoM) [inline src/js/components/Rewards/components/ScrollMenuRewardsItem.js]: looks very similar to `RewardsItem`, just one line less

File: src/styles/rewards/learnGoalSetting.scss [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+      margin: 20px 0;
+      height: fit-content;
+    }
+  }
+}
+
+.teach-badge-earning-item { 
+  display: flex;
+  flex-direction: row;
+align-items: center;
```
Comment (SaenkoM) [inline src/styles/rewards/learnGoalSetting.scss]: indentation

File: src/styles/rewards/rewards.scss [hunk 21]
```diff
@@ -0,0 +1,148 @@
...
+      }
+
+      width: 100%;
+      display: flex;
+      flex-direction: row;
+    }
+  }
+
+  &-item {
+    @media (max-width: 360px) {
```
Comment (SaenkoM) [inline src/styles/rewards/rewards.scss]: usually media tags are used separately at the end of file, so one media per condition, that way you wouldn't miss 360 and 361 difference

File: src/js/components/CheckIn/components/GoalDate/GoalDate.js (hunk 3 above)
Comment (danispavel) [reply src/js/components/CheckIn/components/GoalDate/GoalDate.js]: Logic was written with Date using. I am not sure that replacing on timestamp doesn't break logic

File: src/js/components/Rewards/LearnGoalSetting/components/Step3.js (hunk 15 above)
Comment (danispavel) [reply src/js/components/Rewards/LearnGoalSetting/components/Step3.js]: Only first three steps are similar and maybe they will be changed in future

File: src/js/components/Rewards/components/ScrollMenu.js (hunk 18 above)
Comment (danispavel) [reply src/js/components/Rewards/components/ScrollMenu.js]: It has ```<i>``` inside

File: src/js/components/CheckIn/components/GoalDate/GoalDate.js [hunk 22]
```diff
@@ -92,28 +92,25 @@ class GoalDate extends Component {
...
-      case RUBY_MAX:
-      case DIAMOND_MAX: history.push('/getcheckin')
-        break
-      default: history.push('/dashboard')
-    }
+    if (currentCheckins === BRONZE_MAX
+      || currentCheckins === SILVER_MAX
+      || currentCheckins === GOLD_MAX
+      || currentCheckins === RUBY_MAX
+      || currentCheckins === DIAMOND_MAX) history.push('/getcheckin')
```
Comment (SaenkoM) [inline src/js/components/CheckIn/components/GoalDate/GoalDate.js]: always open braces for `if`, it's easier to see what's going on that way

File: src/styles/rewards/learnGoalSetting.scss [hunk 23]
```diff
@@ -49,19 +50,18 @@
...
-}
-
-&-img-wrapper {
-  width: 20%;
-}
+  align-items: center;
+  justify-content: space-between;
+  width: 100%;
+  padding: 5px 10px;
+  &-text {
```
Comment (SaenkoM) [inline src/styles/rewards/learnGoalSetting.scss]: missing empty line

File: src/js/components/CheckIn/components/GoalDate/GoalDate.js [hunk 24]
```diff
@@ -108,8 +108,11 @@ class GoalDate extends Component {
       || currentCheckins === SILVER_MAX
       || currentCheckins === GOLD_MAX
       || currentCheckins === RUBY_MAX
-      || currentCheckins === DIAMOND_MAX) history.push('/getcheckin')
-    else history.push('/dashboard')
+      || currentCheckins === DIAMOND_MAX) {
```
Comment (SaenkoM) [inline src/js/components/CheckIn/components/GoalDate/GoalDate.js]: can be
```
const validCheckins = [SILVER_MAX, GOLD_MAX...]
if (validCheckins.includes(currentCheckins)) {}
```

**Pull request merged**